### ISSUE TITLE
Proposed changes to the behaviour of the PDF class.

### ISSUE THREAD

**Issue by scarlehoff** [NNPDF/nnpdf#1475]
This is a collection of issues with the current implementation of the PDF class.

This issue will be organized as:
- Problem
- Examples
- Solution

# Error Type

Half of the issues with the PDF class is the management of the error type.

The `nnpdf_error` attribute:

https://github.com/NNPDF/nnpdf/blob/eeae19b15c960b4f8f974c3ac2ba2af36ed845bc/validphys2/src/validphys/core.py#L180

I will remove this attribute and all references to it. If something is accessing this attribute then it is a libNNPDF thing and should be removed. Instead (as was suggested) any behavior that depends on the error type of the PDF should be interfaced through its associated `stats_class`.

# Actual PDF

Currently `load` loads a libNNPDF PDF set.

https://github.com/NNPDF/nnpdf/blob/eeae19b15c960b4f8f974c3ac2ba2af36ed845bc/validphys2/src/validphys/core.py#L162

It will be changed to a python wrapper to the `lhapdf` python interface. It makes sense to do it this way because it allows for certain optimization (such as `grid_values`).

# Attributes

Currently we look inside the info file of the PDF by using `hasattr`, which is confusing and hides what's actually happening.

https://github.com/NNPDF/nnpdf/blob/eeae19b15c960b4f8f974c3ac2ba2af36ed845bc/validphys2/src/validphys/core.py#L101
https://github.com/NNPDF/nnpdf/blob/eeae19b15c960b4f8f974c3ac2ba2af36ed845bc/validphys2/src/validphys/core.py#L149

I will change it to something more explicit.

# Reference to libNNPDF PDF

While there is a few instances where libNNPDF is needed, these are only `t0` situations:

https://github.com/NNPDF/nnpdf/blob/d68954eef2d528c9991e37faa30c2ca477a9208b/validphys2/src/validphys/covmats.py#L476

https://github.com/NNPDF/nnpdf/blob/d68954eef2d528c9991e37faa30c2ca477a9208b/validphys2/src/validphys/covmats.py#L510

https://github.com/NNPDF/nnpdf/blob/d68954eef2d528c9991e37faa30c2ca477a9208b/validphys2/src/validphys/closuretest/multiclosure.py#L36

which is nice because it means I can do all of the above while leaving a `load_nnpdf_t0` interface if needed.

Of course, some of the issues highlighted here might have reasons behind them to which I'm not privy. I might also be missing other points that could be removed and that I'm overlooking.

**Comment by Zaharid** [comment]
I agree with everything, with perhaps some reservation in regards to the last point, the t0 stuff, or the order in which to do various things. I think we could have t0 predictions and t0 covmat in pure python without the need to have to retrofit the cpp functionality in the new pdf, provided the t0 stuff is taken care of first. I suppose it wouldn't matter very much if the cpp interface could be removed easily but it might be a bit cleaner? @siranipour do you agree?

**Comment by scarlehoff** [comment]
The closure tests also need a libNNPDF PDF

https://github.com/NNPDF/nnpdf/blob/eeae19b15c960b4f8f974c3ac2ba2af36ed845bc/validphys2/src/validphys/filters.py#L180

**Comment by siranipour** [comment]
There was a discussion regarding the t0 covmats [here](https://github.com/NNPDF/nnpdf/pull/1433#issuecomment-966994152).  I'm still a tad confused about this point. For example: we have this test that compares a pure python t0 covmat with the C++ t0 covmat:
https://github.com/NNPDF/nnpdf/blob/eeae19b15c960b4f8f974c3ac2ba2af36ed845bc/validphys2/src/validphys/tests/test_covmats.py#L136
Now I agree that
https://github.com/NNPDF/nnpdf/blob/eeae19b15c960b4f8f974c3ac2ba2af36ed845bc/validphys2/src/validphys/config.py#L700
and
https://github.com/NNPDF/nnpdf/blob/eeae19b15c960b4f8f974c3ac2ba2af36ed845bc/validphys2/src/validphys/config.py#L712
still dispatch the old C++ covariance matrix code, but in principle it should be easy to simply swap these out for the Python ones instead no?

In particular, notice that the python t0 covmat code _does not_ use the libNNPDF PDF class, in particular the line here
https://github.com/NNPDF/nnpdf/blob/d68954eef2d528c9991e37faa30c2ca477a9208b/validphys2/src/validphys/covmats.py#L319
is performing the convolution also purely in python. Now, I agree the PDF used in the convolution is C++ spec but surely this is sufficiently well contained that it doesn't directly impact the covmat code?

Again, apologies if I'm missing an obvious point, and for what it's worth I think the `validphys.covmats` module can do with some cleaning, because there is a lot of code that should have been removed long ago, namely the C++ ones (I suppose we kept them for backwards compatibility).

I do, however, agree with the remainder of the points made, and thanks @scarlehoff for a nicely formatted summary of the issues.

**Comment by scarlehoff** [comment]
> it should be easy to simply swap these out for the Python ones instead no?

I don't know. When destroying C++ I would like to avoid removing whole chunks of code just because it is inconvenient, specially if it is being used.

Actually if those two production rules can just be swapped by the python ones I'd appreciate if someone does it so it's one less thing I have to worry about.

Then there's the `MakeClosure` stuff which will still be a problem.

**Comment by scarlehoff** [comment]
Changed the initial comment to remove the two cases that have been removed thanks to #1477

**Comment by scarlehoff** [comment]
@siranipour @Zaharid do you have any feeling on how difficult would it be to change the two places in which we are still using PDFs through libNNPDF? They are both closure-test related. The first one seems -naively- easy to bypass, the second I don't really know.

https://github.com/NNPDF/nnpdf/blob/d68954eef2d528c9991e37faa30c2ca477a9208b/validphys2/src/validphys/closuretest/multiclosure.py#L96
https://github.com/NNPDF/nnpdf/blob/eeae19b15c960b4f8f974c3ac2ba2af36ed845bc/validphys2/src/validphys/filters.py#L180

**Comment by Zaharid** [comment]
@siranipour do we have the capability to to generate closure test replicas with all the options from Python?

**Comment by siranipour** [comment]
Unfortunately, I don't think we do yet... This was something that we failed to remember when implementing pseudodata.py at the time
